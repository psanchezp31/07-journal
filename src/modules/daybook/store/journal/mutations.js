// export const myMutation = (state) =>{
// }

export const setEntries = (state, entries) => {
  state.entries = [...state.entries, ...entries];
  state.isLoading = false;
};
export const updateEntry = (state, updatedEntry) => {
  const idx = state.entries.map((e) => e.id).indexOf(updatedEntry.id);
  state.entries[idx] = updatedEntry
};

export const addEntry = (state, createdEntry) => {
  // state.entries.push(createdEntry)
  state.entries = [createdEntry, ...state.entries] // en el nuevo arreglo, ahora irÃ¡ 
  //la nueva entrada de primero y el resto van despues
};

export const deleteEntry = (state, id) =>{
  state.entries = state.entries.filter(entry => entry.id !== id)
}

export const clearEntries = (state) =>{
  state.entries = []
}
