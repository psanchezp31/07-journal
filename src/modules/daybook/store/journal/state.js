export default () => ({
  isLoading: true,
  entries: [
    {
      id: new Date().getTime(),
      date: new Date().toDateString(),
      text: "Lorem ipsum dolor, sit amet consectetur adipisicing elit. Voluptates,quasi at. Atque, deleniti digniss",
      picture: null,
    },
    {
      id: new Date().getTime() + 1000,
      date: new Date().toDateString(),
      text: "Lorem ipsum dolor, sit amet consectetur adipisicing elit. Voluptates,quasi at. Atque, deleniti digniss",
      picture: null,
    },
    {
      id: new Date().getTime() + 2000,
      date: new Date().toDateString(),
      text: "Lorem ipsum dolor, sit amet consectetur adipisicing elit. Voluptates,quasi at. Atque, deleniti digniss",
      picture: null,
    },
  ],
});
