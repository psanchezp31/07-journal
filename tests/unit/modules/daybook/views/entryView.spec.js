import { shallowMount } from "@vue/test-utils";
import { createStore } from "vuex";
import Swal from "sweetalert2";

import journal from "@/modules/daybook/store/journal";
import EntryView from "@/modules/daybook/views/EntryView.vue";
import { journalState } from "../../../mock-data/test-journal-state";

const createVuexStore = (initialState) =>
  createStore({
    modules: {
      journal: {
        ...journal,
        state: { ...initialState },
      },
    },
  });
//mock de la librería sweetalert2
jest.mock("sweetalert2", () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn(),
}));

describe("pruebas en el EntryView", () => {
  const store = createVuexStore(journalState);
  store.dispatch = jest.fn()  //evitar que el dispatch (o la acción) se ejecute de verdad en la bd, se mockea con un jest.fn
  const mockRouter = {
    push: jest.fn(),
  };
  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks(); //para que limpie las pruebas cada vez que termine cada prueba
    wrapper = shallowMount(EntryView, {
      props: {
        id: "-Mvv17qG_ZsgYZpwMu-X",
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    });
  });

  test("debe de sacar al usuario porque el id no existe", () => {
    const wrapper = shallowMount(EntryView, {
      props: {
        id: "este id no existe en el store",
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    });
    expect(mockRouter.push).toHaveBeenCalledWith({ name: "no-entry" }); //evaluar que al pasarle un id que no existe, lo saque a la vista no-entry
  });

  test("debe de mostrar la entrada correctamente", () => {
    //aca se verifica que como el id montado en en el wrapper del before each existe, muestre la entrada correctamente
    expect(wrapper.html()).toMatchSnapshot();
    expect(mockRouter.push).not.toHaveBeenCalled();
  });

  test("debe de borrar la entrada y salir", (done) => {
    Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true })); //aca fingimos que en la ventana del sweetalert le damos click a si, para eliminar la entrada
    wrapper.find(".btn-danger").trigger("click");
    expect(Swal.fire).toHaveBeenCalledWith({
      title: "¿Está seguro?",
      text: "Una vez borrado, no se puede recuperar",
      showDenyButton: true,
      confirmButtonText: "Sí, estoy seguro",
    });
    setTimeout(() => {
      expect(store.dispatch).toHaveBeenCalledWith("journal/deleteEntry", "-Mvv17qG_ZsgYZpwMu-X")
      expect(mockRouter.push).toHaveBeenCalled();
      done();
    }, 2); //para esperar 1 milisegundo para que se termine de hacer las validaciones en el hilo
  });

});
