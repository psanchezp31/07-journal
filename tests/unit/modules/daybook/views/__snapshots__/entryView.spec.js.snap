// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el EntryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">14</span><span class="mx-1 fs-3">Febrero</span><span class="mx-2 fs-4 fw-light">2022, Lunes</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedió hoy?"></textarea></div><img src="https://res.cloudinary.com/paulis/image/upload/v1644889570/images/j8cgvzq2n14rio1dz94m.jpg" alt="perrito" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
